include/astshim/FrameDict.h

```diff
@@ -0,0 +1,310 @@
+                    This initially becomes both the base and the current Frame.
+                    Further Frames may be added using @ref addFrame
+    @param[in] options  Comma-separated list of attribute assignments.
+    */
+    explicit FrameDict(Frame const &frame, std::string const &options = "") : FrameSet(frame, options) {
+        _rebuildDict();
+    }
+
+    /**
+    Construct a FrameDict from two frames and a mapping that connects them
+
+    Both frames and the mapping are deep copied.
+
+    @param[in] baseFrame  base @ref Frame.
+    @param[in] mapping  mapping connecting baseFrame to currentFrame.
+    @param[in] currentFrame  current @ref Frame.
+    @param[in] options  Comma-separated list of attribute assignments.
+
+    @throws std::invalid_argument if both Frames have the same non-empty domain.
+    */
+    explicit FrameDict(Frame const &baseFrame, Mapping const &mapping, Frame const &currentFrame,
+                       std::string const &options = "")
+            : FrameSet(baseFrame, mapping, currentFrame) {
+        _rebuildDict();
+    }
+
+    /**
+    Construct a FrameDict from a FrameSet
+
+    The FrameSet is deep-copied
+
+    @throws std::invalid_argument if two Frames in the FrameSet have the same non-empty domain.
+    */
+    explicit FrameDict(FrameSet const &frameSet)
+            : FrameDict(reinterpret_cast<AstFrameSet *>(frameSet.copy()->getRawPtr())) {}
+    virtual ~FrameDict() {}
+
+    FrameDict(FrameDict const &) = delete;
+    FrameDict(FrameDict &&) = default;
+    FrameDict &operator=(FrameDict const &) = delete;
+    FrameDict &operator=(FrameDict &&) = default;
+
+    /// Return a deep copy of this object.
+    std::shared_ptr<FrameSet> copy() const { return std::static_pointer_cast<FrameDict>(copyPolymorphic()); }
+
+    void addFrame(int iframe, Mapping const &map, Frame const &frame) override {
+        FrameSet::addFrame(iframe, map, frame);
+        _addFrameToDict(frame, FrameSet::getCurrent());
+    }
+
+    /**
+    Add a Frame connected by a Mapping to an existing initial Frame. A variant of FrameSet::addFrame
+    where the initial frame is specified by domain.
+
+    @throws std::invalid_argument if `frame` has a non-empty domain and this FrameDict already
+    contains a Frame with that domain
+    */
+    void addFrame(std::string const &domain, Mapping const &map, Frame const &frame) {
+        addFrame(getIndex(domain), map, frame);
+    }
+
+    /**
+    Get a list of the domain names for all contained Frames that have non-empty domain names.
+
+    The order of the names is unspecified.
+    */
+    std::vector<std::string> getAllDomains() const {
+        std::vector<std::string> domains;
+        domains.reserve(_domainIndexDict.size());
+        for (auto const &item : _domainIndexDict) {
+            domains.emplace_back(item.first);
+        }
+        return domains;
+    }
+
+    using FrameSet::getFrame;
```

Comment:
This statement is necessary in order for FrameDict to support the FrameSet version of getFrame. Otherwise the domain-name-based variant hides the original version.

If there's a simpler way to do this, I'd love to know about it. I find all the using statements very clumsy.

Reply:
I think you're talking about Effective C++, Item 33. It looks like you can't get rid of the `using` statements without breaking the API, but you *can* remove the explicit qualifiers.

Reply:
My compiler requires the `FrameSet::` in the `using` statements. I also feel it makes it the code a bit clearer.